File: data_preprocessing/lib_external/biosignal/camera/python/heart.py
import cv2
import time
import math
import numpy as np
import mediapipe as mp
from scipy import signal
from argparse import ArgumentParser

import cam


def convert(normalized_x, normalized_y, image_width, image_height):
    x = normalized_x * image_width
    y = normalized_y * image_height
    x_px = max(min(math.floor(x), image_width - 1), 0)
    y_px = max(min(math.floor(y), image_height - 1), 0)
    return x_px, y_px


class Queue:
    # 리스트의 사이즈를 고정하기 위한 클래스입니다.
    def __init__(self, size, value=0.0):
        self.size = size
        self.data = [value for _ in range(self.size)]

    def __len__(self):
        return len(self.data)

    def __call__(self, x):
        self.data.append(x)
        if len(self.data) > self.size:
            self.data.pop(0)

    def __getitem__(self, idx):
        return self.data[idx]


class SignalModule:
    # 심박 신호를 처리하기 위한 클래스입니다.
    def __init__(self, size=360):
        self.size = size
        self.reset()

    def reset(self):
        self.times = Queue(self.size, time.time())
        self.qr = Queue(self.size)
        self.qg = Queue(self.size)
        self.qb = Queue(self.size)
        self.qx = Queue(self.size)
        self.qy = Queue(self.size)
        self.sig = np.zeros(self.size, dtype=float)
        self.sig_filt = np.zeros(self.size, dtype=float)
        self.peaks = []
        self.frequency = 0.0

    def __call__(self, timestamp, r, g, b):
        self.times(timestamp)
        self.qr(r)
        self.qg(g)
        self.qb(b)

        # Convert to PPG signal
        x = 3*r - 2*g
        y = 1.5*r + g - 1.5*b
        self.qx(x)
        self.qy(y)

        x = np.asarray(self.qx.data)
        y = np.asarray(self.qy.data)
        beta = np.std(x) / (np.std(y) + 1e-9)
        s = x - beta*y  # PPG signal

        # TODO: check parameters
        sig = -s
        # Band pass filtering
        filt = signal.firwin(numtaps=31, cutoff=[
                             0.4, 3.0], pass_zero='bandpass', fs=15)
        sig_filt = signal.convolve(sig - sig.mean(), filt, mode='same')
        # Find peaks
        peaks, _ = signal.find_peaks(sig_filt, prominence=0.1)

        self.sig = sig
        self.sig_filt = sig_filt
        self.peaks = peaks

        # calculate frequency
        if len(peaks) >= 2:
            p0 = peaks[0]
            p1 = peaks[-1]
            t0 = self.times[p0]
            t1 = self.times[p1]
            dt = t1 - t0
            count = len(peaks)
            frequency = (count - 1) / (dt - 1e-6)
            self.frequency = frequency

    def get_graph(self, scale=160):
        # 신호 그래프 시각화 부분입니다.
        def minmax_normalize(data):
            return (data - data.min()) / (data.max() - data.min() + 1e-6)

        s = self.sig_filt
        s = (minmax_normalize(s) * scale).astype(int)
        size = len(s)

        def reorder(p0, p1):
            if p0 == p1:
                return p0, p1+1
            elif p0 > p1:
                return p1, p0
            return p0, p1

        graph1 = np.zeros((size, scale, 3), dtype=np.uint8)

        for peak in self.peaks:
            graph1[peak, :, 2] = 255
        for i in range(1, size):
            p0, p1 = reorder(s[i-1], s[i])
            graph1[i, p0:p1, 1] = 255

        v = self.sig
        v = (minmax_normalize(v) * scale).astype(int)
        graph2 = np.zeros((size, scale, 3), dtype=np.uint8)
        for i in range(1, size):
            p0, p1 = reorder(v[i-1], v[i])
            graph2[i, p0:p1, (1, 2)] = 255

        graph = np.concatenate([graph1, graph2], axis=1)

        return graph


def main(recording=False):
    cv2.namedWindow('view')
    cv2.moveWindow('view', 0, 0)
    cv2.namedWindow('hist')
    cv2.moveWindow('hist', 0, 400)
    cv2.namedWindow('crop')
    cv2.moveWindow('crop', 320, 400)
    cv2.namedWindow('mask')
    cv2.moveWindow('mask', 640, 400)

    camera = cam.Cam(0)
    img_w, img_h = 640, 360
    s_module = SignalModule(360)
    mp_face_detection = mp.solutions.face_detection

    if recording:
        rec = cam.Rec('./records', 15.0, (960, 360))

    with mp_face_detection.FaceDetection(model_selection=0, min_detection_confidence=0.5) as face_detection:
        while camera():
            image = camera.frame
            timestamp = camera.time

            """
            Preprocess
            이미지를 적당한 사이즈로 리사이즈합니다.
            연산량을 고려하여 사이즈를 설정합니다.
            """
            image = cv2.resize(image, (img_w, img_h))

            """
            Detect
            https://google.github.io/mediapipe/solutions/face_detection
            mediapipe를 사용하여 face detection을 합니다.
            OpenCV에서는 기본적으로 BGR 순서의 색채널을 사용하기 때문에
            RGB 채널로 변환 후 처리해야 합니다.
            """
            image.flags.writeable = False
            image = cv2.cvtColor(image, cv2.COLOR_BGR2RGB)
            results = face_detection.process(image)
            image = cv2.cvtColor(image, cv2.COLOR_RGB2BGR)
            image.flags.writeable = True

            """
            Convert
            검출한 bbox의 좌표를 변환합니다.
            출력된 결과는 normalized되어 있어 이미지 사이즈만큼 scaling해야합니다.
            p0 -> top left corner point
            p1 -> bottom right corner point
            """
            bboxes = []
            detections = results.detections
            if detections:
                for detection in detections:
                    location = detection.location_data
                    rel_bb = location.relative_bounding_box

                    p0 = convert(rel_bb.xmin, rel_bb.ymin, img_w, img_h)
                    p1 = convert(rel_bb.xmin+rel_bb.width,
                                 rel_bb.ymin+rel_bb.height, img_w, img_h)

                    bbox = (p0, p1)
                    bboxes.append(bbox)

            if len(bboxes):
                bbox = bboxes[0]
                # 이미지에서 bbox부분을 crop합니다.
                (x0, y0), (x1, y1) = bboxes[0]
                crop = image[y0:y1, x0:x1].copy()

                # 이미지를 HSV로 변환하고 S 채널에 대한 히스토그램을 구합니다.
                hsv = cv2.cvtColor(crop, cv2.COLOR_BGR2HSV)
                hist = cv2.calcHist([hsv], [1], None, [256], [0, 256])
                hist = hist.squeeze()
                hist = signal.medfilt(hist, 5)

                # S 히스토그램에서의 최대값을 기준으로 S 값의 범위를 정합니다.
                min_pixel = 32
                hist[:min_pixel] = 0
                hist_max = np.argmax(hist)
                alpha = 0.4
                TH_range = alpha * hist_max
                range0 = int(hist_max - TH_range/2)
                range1 = int(hist_max + TH_range/2)

                sat = hsv[:, :, 1]
                mask = (sat > range0) & (sat < range1)

                # mask에 해당하는 픽셀들을 모아 R,G,B 각각의 평균값을 계산합니다.
                pixels = crop[mask]
                mean = pixels.sum(axis=0) / (len(pixels) + 1e-6)
                b, g, r = mean
                # R,G,B 평균값을 타임스탬프와 함께 signal module에 입력합니다.
                s_module(timestamp, r, g, b)

                # S 히스토그램 시각화 부분입니다.
                scale = 320
                value = np.clip(hist, 0, scale).astype(int)
                hist_view = np.zeros((256, scale, 3), np.uint8)
                for i in range(0, 256):
                    hist_view[i, :value[i], 1] = 64
                hist_view[hist_max, :, 2] = 255
                hist_view[range0, :, 0] = 255
                hist_view[range1, :, 0] = 255

                cv2.imshow('crop', crop)
                mask_view = mask.reshape(
                    mask.shape[0], mask.shape[1], 1) * crop
                cv2.imshow('mask', mask_view)
                cv2.imshow('hist', hist_view)

            # Visualize
            for p0, p1 in bboxes:
                cv2.rectangle(image, p0, p1, (0, 255, 0), 1)
            logs = [
                ' %4.1f FPS' % camera.fps,
                ' %4.1f bpm' % (s_module.frequency * 60)
            ]
            image[:50, :100] = 0
            color = (0, 255, 255)
            org = [0, 20]
            for log in logs:
                cv2.putText(image, log, org, 0, 0.5, color, 1)
                org[1] += 20
            graph = s_module.get_graph()
            view = np.concatenate([image, graph], axis=1)
            cv2.imshow('view', view)
            if recording:
                rec(view)


if __name__ == '__main__':
    parser = ArgumentParser()
    parser.add_argument('--recording', type=int, default=0)
    args = parser.parse_args()
    main(recording=args.recording)

File: data_preprocessing/lib_external/biosignal/camera/cpp/app/heart.cc
#include <algorithm>
#include <cmath>
#include <cstdio>
#include <iostream>
#include <vector>

#include <opencv2/opencv.hpp>

#include "cam.h"
#include "signalgo.h"

class Queue {
  unsigned int size_;

public:
  std::vector<double> data;
  Queue();
  Queue(int size);
  void set(int size);
  void push_back(double x);
  int size() { return size_; };
  double mean();
  double var();
  double std();
};

Queue::Queue() { size_ = 0; }

void Queue::set(int size) {
  size_ = size;
  for (int i = 0; i < size; i++) {
    data.push_back(0.0);
  }
}

void Queue::push_back(double x) {
  data.push_back(x);
  if (data.size() > size_) {
    data.erase(data.begin());
  }
}

double Queue::mean() {
  if (data.size() == 0)
    return 0.0;
  double sum = 0;
  for (int i = 0; i < data.size(); i++) {
    sum += data[i];
  }
  return sum / data.size();
}

double Queue::var() {
  if (data.size() == 0)
    return 0.0;
  double m = mean();
  double sum = 0;
  for (int i = 0; i < data.size(); i++) {
    double d = data[i] - m;
    sum += (d * d);
  }
  return sum / data.size();
}

double Queue::std() { return sqrt(var()); }

double clip(double value, double min, double max) {
  if (value < min) {
    return min;
  } else if (value > max) {
    return max;
  }
  return value;
}

double *bandpassfilter(double fl, double fh, double fs, int N) {
  double *filter = NULL;
  double flc = 0;
  double omegal = 0;
  double fhc = 0;
  double omegah = 0;
  int i = 0;
  int middle = 0;

  if (fl <= 0 || fh <= 0 || fl >= fh || fs <= 0 || N <= 0)
    return NULL;

  filter = new double[N];
  if (!filter)
    return NULL;

  memset(filter, 0, N * sizeof(double));

  middle = (int)(N / 2);
  flc = fl / fs;
  omegal = 2 * M_PI * flc;
  fhc = fh / fs;
  omegah = 2 * M_PI * fhc;

  for (i = 0; i < N; i++) {
    if (i == middle) {
      filter[i] = 2 * fhc - 2 * flc;
    } else {
      filter[i] = sin(omegah * (i - middle)) / (M_PI * (i - middle)) -
                  sin(omegal * (i - middle)) / (M_PI * (i - middle));
    }
  }
  return filter;
}

class SignalModule {
  int size_;

public:
  Queue times;
  Queue qx;
  Queue qy;
  double *sig;
  double *sig_filt;
  int filter_size;
  double *filter;
  std::vector<int> peaks;
  double frequency;
  SignalModule(int size);
  ~SignalModule();
  void run(double timestamp, double r, double g, double b);
  cv::Mat get_graph();
};

SignalModule::SignalModule(int size) {
  size_ = size;
  times.set(size);
  qx.set(size);
  qy.set(size);
  sig = new double[size];
  sig_filt = new double[size];
  filter_size = 33;
  filter = bandpassfilter(0.4, 2.0, 15, filter_size);
  frequency = 0.0;
}

SignalModule::~SignalModule() {
  delete[] sig;
  delete[] filter;
}

void SignalModule::run(double timestamp, double r, double g, double b) {
  double x = 3 * r - 2 * g;
  double y = 1.5 * r + g - 1.5 * b;
  qx.push_back(x);
  qy.push_back(y);

  double beta = qx.std() / (qy.std() + 1e-9);

  for (int i = 0; i < size_; i++) {
    sig[i] = beta * qx.data[i] - qy.data[i];
  }

  // filtering
  for (int i = filter_size; i < size_; i++) {
    double value = 0;
    for (int j = 0; j < filter_size; j++) {
      value += (sig[i - j] * filter[j]);
    }
    sig_filt[i] = value;
  }

  std::vector<double> src(sig_filt, sig_filt + size_);

  peaks = signalgo::find_peaks(src, 0.1);
}

cv::Mat SignalModule::get_graph() {
  int scale = 160;
  cv::Mat graph1 = cv::Mat::zeros(cv::Size(scale, size_), CV_8UC3);
  double *s = sig;
  double min = s[0];
  double max = s[0];
  for (int i = 1; i < size_; i++) {
    if (s[i] > max) {
      max = s[i];
    }
    if (s[i] < min) {
      min = s[i];
    }
  }
  double max_min = max - min + 1e-6;

  int s_prev = (int)((s[0] - min) / max_min * scale);
  for (int i = 1; i < size_; i++) {
    double v = s[i];
    v = (v - min) / max_min;
    int s = (int)(v * scale);

    int p0 = std::min(s, s_prev);
    int p1 = std::max(s, s_prev);
    for (int p = p0; p < p1; p++) {
      graph1.data[(i * scale + p) * 3 + 1] = 255; // green channel
    }
    s_prev = s;
  }

  cv::Mat graph2 = cv::Mat::zeros(cv::Size(scale, size_), CV_8UC3);
  s = sig_filt;
  min = s[filter_size];
  max = s[filter_size];
  for (int i = filter_size + 1; i < size_; i++) {
    if (s[i] > max) {
      max = s[i];
    }
    if (s[i] < min) {
      min = s[i];
    }
  }
  max_min = max - min + 1e-6;

  s_prev = (int)((s[filter_size] - min) / max_min * scale);
  for (int i = filter_size + 1; i < size_; i++) {
    double v = s[i];
    v = (v - min) / max_min;
    int s = (int)(v * scale);

    int p0 = std::min(s, s_prev);
    int p1 = std::max(s, s_prev);
    for (int p = p0; p < p1; p++) {
      graph2.data[(i * scale + p) * 3 + 1] = 255; // green channel
    }
    s_prev = s;
  }

  for (int i = 0; i < peaks.size(); i++) {
    int peak = peaks[i];
    for (int j = 0; j < scale; j++) {
      graph2.data[3 * (peak * scale + j) + 2] = 255;
    }
  }

  cv::Mat graph;
  cv::hconcat(graph1, graph2, graph);

  return graph;
}

void filtering(int size, double *samples) {}

int test() {
  cv::namedWindow("view");
  cv::moveWindow("view", 0, 0);
  cv::namedWindow("crop");
  cv::moveWindow("crop", 0, 400);
  cv::namedWindow("mask");
  cv::moveWindow("mask", 0, 560);
  cv::namedWindow("hist");
  cv::moveWindow("hist", 320, 400);

  cam::Cam camera(0);
  int width = camera.get_width() / 2;
  int height = camera.get_height() / 2;

  auto s_module = SignalModule(360);

  cv::Mat image;

  while (camera.read()) {
    auto frame = camera.get_frame();
    cv::resize(frame, image, cv::Size(width, height));

    int bsize = 120;
    cv::Rect bbox((width - bsize) / 2, (height - bsize) / 2, bsize, bsize);
    cv::Mat crop = image(bbox).clone();

    cv::Mat hsv;
    cv::cvtColor(crop, hsv, cv::COLOR_BGR2HSV);

    std::vector<cv::Mat> hsv_planes;
    cv::split(hsv, hsv_planes);
    // cv::imshow("hsv", hsv_planes[1]);

    int hist_size = 256;
    float range[] = {0, 256};
    const float *hist_range[] = {range};
    bool uniform = true, accumulate = false;
    cv::Mat hist, temp;
    cv::calcHist(&hsv_planes[1], 1, 0, temp, hist, 1, &hist_size, hist_range,
                 uniform, accumulate);

    cv::Mat hist_m;
    cv::medianBlur(hist, hist_m, 5);

    double minVal;
    double maxVal;
    cv::Point minLoc;
    cv::Point maxLoc;

    cv::minMaxLoc(hist_m, &minVal, &maxVal, &minLoc, &maxLoc);
    int hist_max = maxLoc.y;

    double alpha = 0.4;
    double TH_range = alpha * hist_max;
    int range0 = (int)(hist_max - TH_range / 2);
    int range1 = (int)(hist_max + TH_range / 2);

    cv::Mat sat = hsv_planes[1];
    int sat_size = sat.rows * sat.cols;

    std::vector<int> mask;
    for (int i = 0; i < sat_size; i++) {
      if (sat.data[i] > range0) {
        if (sat.data[i] < range1) {
          mask.push_back(i);
        }
      }
    }

    double r_sum = 0, g_sum = 0, b_sum = 0;
    for (int i = 0; i < mask.size(); i++) {
      int index = mask[i];
      b_sum += crop.data[3 * index];
      g_sum += crop.data[3 * index + 1];
      r_sum += crop.data[3 * index + 2];
    }
    double r = r_sum / (mask.size() + 1e-9);
    double g = g_sum / (mask.size() + 1e-9);
    double b = b_sum / (mask.size() + 1e-9);

    s_module.run(camera.get_time(), r, g, b);

    int scale = 320;
    cv::Mat hist_view = cv::Mat::zeros(cv::Size(scale, 256), CV_8UC3);
    for (int i = 0; i < 256; i++) {
      int value = (int)hist_m.at<float>(i);
      if (value >= scale)
        value = scale - 1;
      for (int j = 0; j < value; j++) {
        hist_view.data[3 * (i * scale + j) + 1] = 64;
      }
    }
    for (int j = 0; j < scale; j++) {
      hist_view.data[3 * (hist_max * scale + j) + 2] = 255;
      hist_view.data[3 * (range0 * scale + j) + 0] = 255;
      hist_view.data[3 * (range1 * scale + j) + 0] = 255;
    }

    cv::Mat mask_view = cv::Mat::zeros(crop.size(), CV_8UC3);
    for (int i = 0; i < mask.size(); i++) {
      int index = 3 * mask[i];
      mask_view.data[index + 0] = crop.data[index + 0];
      mask_view.data[index + 1] = crop.data[index + 1];
      mask_view.data[index + 2] = crop.data[index + 2];
    }

    cv::Mat graph = s_module.get_graph();

    cv::rectangle(image, bbox, cv::Scalar(0, 255, 0), 1);
    cv::Mat view;
    cv::hconcat(image, graph, view);

    cv::imshow("view", view);
    cv::imshow("crop", crop);
    cv::imshow("hist", hist_view);
    cv::imshow("mask", mask_view);
  }

  return 0;
}

int main() {
  test();
  return 0;
}